app/models/user.rb

```diff
@@ -926,6 +927,15 @@ def password_validator
     PasswordValidator.new(attributes: :password).validate_each(self, :password, @raw_password)
   end
 
+  def password_expired?(password)
+    passwords
```

Comment:
Okay so the main problem I see with this right now is that, if the user's old password fits the password requirements set out by admins, they can set their password to the same thing as the old one in the password reset flow.

Then, when they log in again, they hit the "Your current password does not meet new password policy requirements for this site. Please check your email for password reset instructions." message.

Perhaps we could also add a check into the password reset flow to say "You cannot set your new password to a previous expired password"?

We may also want to clean up these expired passwords at some point, we don't want to get to a point where the user is forced to choose passwords they don't remember (yes a password manager is the ideal solution of course)

Reply:
Yeah we need to ban then from setting the password to the expired password in the forget password flow... should be easy no?

Reply:
I think it is here? right: https://github.com/discourse/discourse/blob/62b1b69e820bfa48898d5f654c2b71162078a73a/app/controllers/users_controller.rb#L885-L886

Reply:
We have a `PasswordValidator` so I can add a conditional there
 
https://github.com/discourse/discourse/blob/30f55cd64b4a1aadab234c63452685ed47a23d35/lib/validators/password_validator.rb#L20-L21

Reply:
This is done in https://github.com/discourse/discourse/pull/27314/files#diff-b2cba4f5709e06b1e090353a85086a5d7e3b81bb4e712b7e87ad8e11d073e95bR22-R23